rosidl_generator_c/rosidl_generator_c/__init__.py

```diff
@@ -13,77 +13,85 @@
 from rosidl_cmake import read_generator_arguments
-from rosidl_parser import parse_message_file
-from rosidl_parser import parse_service_file
+from rosidl_parser.definition import AbstractType
+from rosidl_parser.definition import Array
+from rosidl_parser.definition import BaseString
+from rosidl_parser.definition import BasicType
+from rosidl_parser.definition import IdlLocator
+from rosidl_parser.definition import NamespacedType
+from rosidl_parser.definition import Sequence
+from rosidl_parser.definition import String
+from rosidl_parser.definition import WString
+from rosidl_parser.parser import parse_idl_file
 
 
 def generate_c(generator_arguments_file):
     args = read_generator_arguments(generator_arguments_file)
 
     template_dir = args['template_dir']
-    mapping_msgs = {
-        os.path.join(template_dir, 'msg.h.em'): '%s.h',
-        os.path.join(template_dir, 'msg__functions.c.em'): '%s__functions.c',
-        os.path.join(template_dir, 'msg__functions.h.em'): '%s__functions.h',
-        os.path.join(template_dir, 'msg__struct.h.em'): '%s__struct.h',
-        os.path.join(template_dir, 'msg__type_support.h.em'): '%s__type_support.h',
+    mapping = {
+        os.path.join(template_dir, 'idl.h.em'): '%s.h',
+        os.path.join(template_dir, 'idl__functions.c.em'): '%s__functions.c',
+        os.path.join(template_dir, 'idl__functions.h.em'): '%s__functions.h',
+        os.path.join(template_dir, 'idl__struct.h.em'): '%s__struct.h',
+        os.path.join(template_dir, 'idl__type_support.h.em'): '%s__type_support.h',
     }
-    mapping_srvs = {
-        os.path.join(template_dir, 'srv.h.em'): '%s.h',
-    }
-    for template_file in list(mapping_msgs.keys()) + list(mapping_srvs.keys()):
+    for template_file in mapping.keys():
         assert os.path.exists(template_file), 'Could not find template: ' + template_file
 
     functions = {
         'get_header_filename_from_msg_name': convert_camel_case_to_lower_case_underscore,
     }
     latest_target_timestamp = get_newest_modification_time(args['target_dependencies'])
 
-    for ros_interface_file in args['ros_interface_files']:
-        extension = os.path.splitext(ros_interface_file)[1]
-        subfolder = os.path.basename(os.path.dirname(ros_interface_file))
-        if extension == '.msg':
-            spec = parse_message_file(args['package_name'], ros_interface_file)
-            for template_file, generated_filename in mapping_msgs.items():
+    for idl_tuple in args.get('idl_tuples', []):
+        idl_parts = idl_tuple.split(':', 1)
+        assert len(idl_parts) == 2
+        locator = IdlLocator(*idl_parts)
+        idl_rel_path = pathlib.Path(idl_parts[1])
+        try:
+            idl_file = parse_idl_file(
+                locator, png_file=os.path.join(
+                    args['output_dir'], str(idl_rel_path.parent),
+                    idl_rel_path.stem) + '.png')
+            for template_file, generated_filename in mapping.items():
```

Comment:
> > I'm wondering if it would be possible to add a step to the pipeline where an `<action>.idl` can be converted to an `<action_plus_generated_messages_and_services>.idl`. 

> Instead, would it be possible to add a step after `rosidl_adapter` but before calling the generators that parses an `<action>.idl` and creates a  `<send goal service>.idl`, `<get result service>.idl` and `<feedback message>.idl`?

I don't think these files need to be generated explicitly. When an `.action` file is parsed (actually the equivalent `.idl` file) the object representation can provide these *extended* derived definitions. So each generator handling an action can feed these sub-parts to an existing message / service template. Similar to how a service objects contains two message objects and just invoked the message template twice - once for the request, once for the response - and optionally generates something service specific in addition. The action object would just contain not only plain messages / services which have 1-to-1 the same content as in the `.action` file but contain the "enriched" versions of the derived types.